### DISCUSSION TITLE
Using Real Applications in a Simulated Network - Showcase

### DISCUSSION THREAD

**Discussion by adamgeorge309** [inet-framework/inet#629] [Showcases]
Discuss the "Using Real Applications in a Simulated Network" showcase here. Ideas and recommendations are welcome.

**Comment by yan711** [comment]
Hi Adam,

I am a beginner of using OMNET++ as well as INET. I am following this showcase to run real applications in a simulated network. However, i met two errors and the packet cannot be transported by the router to tapb. The first one is "Transport prototol UDP is not connected", then i added "*.router.hasUDP = true" in the .ini, and the problem was solved. The second one is "No socket registered on port 4004", and i have no idea how to solve it. Isn't "vlc rtp://192.168.3.20:4004 &" supposed to bind the socket on the port 4004?
Do you know how to solve this issue?
Btw, i am using OMNET++5.5.1 and INET 4.2.0.
Thanks in advance!

**Comment by adamgeorge309** [comment]
Can you try it with the omnetpp master branch and INET 4.2? Also, the video file in the script is not in the repo for now, so use another one and replace the filename in the script. Thanks for reporting this.

**Comment by yan711** [comment]
Thank you for your answer! I installed omnetpp master branch with a failure, so I tried with OMNeT++5.6.2 and 5.6.1 with INET 4.2. The same two errors occured. And I wrote a udp socket which is binded on "192.168.3.20" and port 4004 in real application, and got the same errors.
Do you know what might cause this error?

**Comment by AAGiron** [comment]
Hi,

I'm executing OMNET 6.0 and INET 4.4. First I've got a syntax error in `*.router.ipv4.natTable.config`, fixed by escaping an '&'. Then, when I run it returns:
```<!> Error: Gate VideostreamingShowcaseNetwork.host1.eth[0].mac.upperLayerIn is not connected to a module of type queueing::IPacketQueue -- in module (inet::EthernetMac) VideostreamingShowcaseNetwork.host1.eth[0].mac (id=206), during network initialization```

Any suggestion?

Thanks in advance

**Reply by AAGiron** [reply]
Hi,

For some reason, using 'setcap' didn't work for me, but I fixed it by executing inet as root user (modifying `run.sh` script).

**Reply by faheemkhuhawar** [reply]
How did you resolve this error? 

`<!> Error: Gate VideoStreamingShowcase.host1.eth[0].mac.upperLayerIn is not connected to a module of type queueing::IPacketQueue -- in module (inet::EthernetMac) VideoStreamingShowcase.host1.eth[0].mac (id=206), during network initialization`

And this error,

`[00007f1940002580] main filter error: Failed to create video converter`

Please guide.

I modified the following lines in `omnetpp.ini` to be as,

```
*.router.ipv4.natTable.config = xml("<config> \
                                        <entry type='prerouting' \
                                        packetFilter='has(Ipv4Header) Ipv4Header.destAddress == \"192.168.2.99\"' \
                                        srcAddress='192.168.3.99' destAddress='192.168.3.20'/> \
                                     </config>")
```

**Reply by adamgeorge309** [reply]
This error is resolved in the master branch.

**Reply by tortoiseme** [reply]
Hi, 

I am getting this error: main filter error: Failed to create video converter. How did you resolve it, please?

**Comment by chivas1000** [comment]
@adamgeorge309 
Hi, I'm using ubuntu18.04 and want to try this videostreaming case, because of the ubuntu version I installed omnetpp 5.6.2 and INET 4.2.5 since only this can run properly, but in the INET github repository I didn't find any videosteaming case in TAG before INET4.3.0.

I copied the videostreaming showcase in my INET tried to run it and failed too.

would you mind tell me which would be the right version of omnetpp and INET for ubuntu18.04 or is there any ways to run this in 18.04?

Thanks.

**Comment by xingyeLiu123** [comment]
Hi, I am wondering why vlc puts its data stream into Router's IP(192.168.2.99) rather than Tapa's IP(192.168.2.20)? When I use a socket to send some UDP packets to the Tapa's IP(192.168.2.20), the emulation network could not recognize these packets, however, when I change the destination IP to 192.168.2.99, the emulation network received these UDP packets.

**Reply by xingyeLiu123** [reply]
Besides, when I add a new TAP device(StandardHost module) into emulation network with a new IP 192.168.2.30 and copy its configuration form tapa, which named tapc, and create a socket bind with IP 192.168.2.30. It is strange that when I use this socket to send UDP packets to Router's IP(192.168.2.99), tapa recognize these UDP packets instead of tapc, since the socket is bound with tapc's IP address. This way, I am also wondering whether there is any hidden binding mechanism in the videostreaming showcase, which restricts adding more TAP devices into emulation network. If there isn't, how can I add new TAP devices and use them to receive the data stream form external application? I am looking forward to your reply. Thank you very much!

**Reply by xingyeLiu123** [reply]
Supplementary explanation: StandardHost module contains ExtUpperEthernetInterface and I have already bind these StandardHost modules with different TAP devices(i.e. tapa, tapc) through their device(string) parameter.

**Reply by adamgeorge309** [reply]
To address the first question, if the packets are sent to 192.168.3.20 (the destination), the packets wouldn't go through the simulated network, but would be sent via the loopback interface. So the packets are sent to the router's interface, where the destination address is rewritten to 192.168.3.20.

**Reply by adamgeorge309** [reply]
To the second question, can you be more specific about what you did? It should be possible to add other hosts/tap interfaces.

**Reply by xingyeLiu123** [reply]
Thanks for your reply！Here is what I have done: I have tried to create 6 sockets for transmitting emulation, which are treated as real applications, and I have bound these 6 sockets with socket.bind() function to 6 TAP devices through IP address. Then, I have created 6 StandardHost modules in simulated network, and each of them is bound with a TAP device using "*.hostx.eth[0].device = "tapx". Here is what I want to do: I want these StandardHost with ExtUpperEthernetInterface(contains TAP device) to represent the above 6 real applications(sockets) in simulated network. And the "_Mapping Relationship_" between the real socket applications and the StandardHost modules is achieved by using socket.sendto(UDP data,"tapx's IP address") function. Then, the UDP packets can enter the simulated network through above mentioned "_Mapping Relationship_". Moreover, the transmission between these real socket applications can be simulated in OMNeT++, thus achieving performance statistics. To summarize, the original transmission between real applications is **Socket1-->Socket2**, I am trying to change it to **Socket1(Real)-->Tap1(StandardHost1)(Simulated)-->Router(Simulated)-->Tap2(StandardHost2)(Simulated)-->Socket2(Real)** by using emluation. Can you provide some suggestions? If you have any question about my statement, please point it out. Thank you very much!

**Reply by xingyeLiu123** [reply]
In a word, I am wondering whether the StandardHost module can receive the UDP packets form a real application through TAP device rather than the Router module, and further achieving packet transmission in simulated network.
